concordium-consensus/src/Concordium/KonsensusV1/Types.hs

```diff
@@ -0,0 +1,804 @@
+-- types are allowed in the block.
+getBakedBlock :: SProtocolVersion pv -> TransactionTime -> Get BakedBlock
+getBakedBlock spv tt = label "BakedBlock" $ do
+    bbRound <- get
+    bbEpoch <- get
+    bbTimestamp <- get
+    bbBaker <- get
+    bbBakerKey <- get
+    bbQuorumCertificate <- get
+    BakedBlockFlags{..} <- get
+    bbTimeoutCertificate <-
+        if bbfTimeoutCertificate
+            then Present <$> get
+            else return Absent
+    bbEpochFinalizationEntry <-
+        if bbfEpochFinalizationEntry
+            then Present <$> get
+            else return Absent
+    bbNonce <- get
+    bbStateHash <- get
+    bbTransactionOutcomesHash <- get
+    numTrans <- getWord64be
+    -- We check that there is at least one byte remaining in the serialization per transaction.
+    -- This is to prevent a malformed block from causing us to allocate an excessively large vector,
+    -- as this could lead to an out-of-memory error.
+    remBytes <- remaining
+    when (fromIntegral remBytes < numTrans) $
+        fail $
+            "Block should have "
+                ++ show numTrans
+                ++ " transactions, but only "
+                ++ show remBytes
+                ++ " bytes remain"
+    bbTransactions <- Vector.replicateM (fromIntegral numTrans) (getBlockItemV0 spv tt)
+    return BakedBlock{..}
+
+-- |A baked block, together with the block hash and block signature.
+--
+-- Invariant: @sbHash == getHash sbBlock@.
+data SignedBlock = SignedBlock
+    { -- |The block contents.
+      sbBlock :: !BakedBlock,
+      -- |The hash of the block.
+      sbHash :: !BlockHash,
+      -- |Signature of the baker on the block.
+      sbSignature :: !BlockSignature
+    }
+
+instance BakedBlockData SignedBlock where
+    blockQuorumCertificate = bbQuorumCertificate . sbBlock
+    blockBaker = bbBaker . sbBlock
+    blockBakerKey = bbBakerKey . sbBlock
+    blockTimeoutCertificate = bbTimeoutCertificate . sbBlock
+    blockEpochFinalizationEntry = bbEpochFinalizationEntry . sbBlock
+    blockNonce = bbNonce . sbBlock
+    blockSignature = sbSignature
+
+instance HashableTo BlockHash SignedBlock where
+    getHash = sbHash
+
+instance Monad m => MHashableTo m BlockHash SignedBlock
+
+-- |Serialize a 'SignedBlock', including the signature.
+putSignedBlock :: Putter SignedBlock
+putSignedBlock SignedBlock{..} = do
+    putBakedBlock sbBlock
+    put sbSignature
+
+-- |Deserialize a 'SignedBlock'. The protocol version is used to determine which transactions types
+-- are permitted.
+getSignedBlock :: SProtocolVersion pv -> TransactionTime -> Get SignedBlock
+getSignedBlock spv tt = do
+    sbBlock <- getBakedBlock spv tt
+    let sbHash = getHash sbBlock
+    sbSignature <- get
+    return SignedBlock{..}
+
+-- |The bytes that are signed by a block signature.
+blockSignatureMessageBytes :: BlockHash -> BS.ByteString
+blockSignatureMessageBytes = Hash.hashToByteString . blockHash
```

Comment:
We could also use domain separators here since baker keys are used to sign different kinds of messages.

Reply:
Will add a task to review choice of domain separators.